### PR TITLE
Use JsonApi in routes to register adapters

### PR DESCRIPTION
**Pull request by pedroluislopez**
This feature allow register adapters in the routes PHP file when you define JSON API resources. This feature avoid use a configuration file for register adapters and/or Eloquent adapter mappings. Example:

```
Route::group([
    'middleware' => ['api-v1'],
    'namespace' => 'Api',
    'prefix' => 'api/v1',
    'as' => 'api-v1::'
], function () {
    JsonApi::eloquentAdapter('comments', 'App\Comment');
    JsonApi::eloquentAdapter('people', 'App\Person');
    JsonApi::eloquentAdapter('posts', 'App\Post');
    JsonApi::adapter('sites', 'App\JsonApi\Sites\Adapter');
});
```

### CODE REVIEW

File: src/Adapters/EloquentAdapter.php [hunk 1]
```diff
@@ -161,4 +161,20 @@ protected function resolveQualifiedKeyName(Model $model, $resourceType)
...
     }
+
+    /**
+     * Add a resource type and key name to Eloquent adapter
+     *
+     * @param string $resourceType
+     * @param string $model
+     * @param string $keyName
+     */
+    public function addResource($resourceType, $model, $keyName = NULL)
```
Comment (jstoone) [inline src/Adapters/EloquentAdapter.php]: Use lowercase `null` in default param value

```php
public function addResource($resourceType, $model, $keyName = null)
```

File: src/Adapters/EloquentAdapter.php [hunk 2]
```diff
@@ -161,4 +161,20 @@ protected function resolveQualifiedKeyName(Model $model, $resourceType)
...
+    {
+        $this->map[$resourceType] = $model;
```
Comment (jstoone) [inline src/Adapters/EloquentAdapter.php]: Add line between this statement and `if`-statement.
```php
$this->map[$resourceType] = $model;

if (! is_null($keyName)) {
    $this->keyNames[$resourceType] = $keyName;
}
```

File: src/Services/JsonApiService.php [hunk 3]
```diff
@@ -67,6 +69,50 @@ public function resource($resourceType, $controller = null, array $options = [])
...
+     * @param string|null $controller
+     * @param array $options
+     * @return ResourceRegistrar
+     */
+    public function adapter($resourceType, $adapter, $controller = null, array $options = [])
+    {
+        $adapterObject = $this->container->make($adapter);
+        
+        // Check if adapter recognises resource type
+        if(! $adapterObject->recognises($resourceType)) {
```
Comment (jstoone) [inline src/Services/JsonApiService.php]: Add space between `if` and the start paran `(`:
```php
if (! $adapterObject->recognises($resourceType)) {
```

File: src/Services/JsonApiService.php [hunk 4]
```diff
@@ -67,6 +69,50 @@ public function resource($resourceType, $controller = null, array $options = [])
...
+     * Register a resource type with the router and include to the Eloquent adapter.
+     *
+     * @param string $resourceType
+     * @param string $model
+     * @param string $keyName
+     * @param string|null $controller
+     * @param array $options
+     * @return ResourceRegistrar
+     */
+    public function eloquentAdapter($resourceType, $model, $keyName = NULL, $controller = null, array $options = [])
```
Comment (jstoone) [inline src/Services/JsonApiService.php]: Make the default `null` parameter lowercase:
```php
public function eloquentAdapter($resourceType, $model, $keyName = null, $controller = null, array $options = [])
```

File: src/Adapters/EloquentAdapter.php (hunk 2 above)
Comment (jstoone) [inline src/Adapters/EloquentAdapter.php]: Also note the extra space between `if` and the start paran `(`.

**Comment by pedroluislopez** [comment]
Done :)

**Comment by jstoone** [comment]
Nice @pedroluislopez! 💯

**Comment by lindyhopchris** [comment]
Sorry about this, but I don't think this will work if you cache your routes. Have you tried it with cached routes?

Also, I'm not sure it solves any problem. You say that it avoids putting stuff in configuration, but the config is there for a reason.

**Comment by pedroluislopez** [comment]
Effectively it doesn't work with `cache: routes`.

About the purpose of this PR:

The modern frameworks do implement automatic features in order to avoid complex configuration files that are modified, usually, by several developers. For example, we work in the same project five developers.

Laravel has this philosophy. In fact, you use this philosophy in two aspects:

1. You don't need to define in which folder the controller are. You define a namespace in routes.php for controllers and the framework search the controllers in the folder `app/Controllers/{namespace}/...`. The name also is calculate in `CloudCreativity\LaravelJsonApi\Routing\ResourceRegistrar` class.

```
    /**
     * @param $resourceType
     * @return string
     */
    protected function controllerFor($resourceType)
    {
        return Str::studly($resourceType) . 'Controller';
    }
```

2. Also, you define a schema where you create Schemas and Adapters PHP files. It's used by `php artisan` commands. It's part of Laravel philosophy.

```
'generator' => [
        'namespace' => 'JsonApi',
        'by-resource' => true,
        'use-eloquent' => true,
    ],
```

However, effectively this solution doesn't work. If you like, I can offer other solutions based on the schema that you define for `php artisan` commands (always leaving a chance to be modify by developers in `json-api` configuration file).

I will close this PR, we will do other PR with other solutions and discuss about them if you like.

**Comment by lindyhopchris** [comment]
Wherever we can move to automation, within the ways that Laravel 5 provides, I am supportive of it. The thing to remember is that this package is still in development (hence no `1.0` tag yet), and whatever your proposing needs to fit in with the general direction of where the package needs to head.

Please remember that in any automation you are proposing it must be capable of supporting multiple API definitions mapping to one set of models. I.e. the package has to support having multiple APIs that expose models in different ways. 

I'd also strongly suggest you submit a proposal for the change via an issue before writing any code. That way we can discuss the idea before any code is written.

We are already working on changes to this package, which I've written up here so you can take a look:
https://github.com/cloudcreativity/laravel-json-api/issues/53

**Pull request closed without merging**